# Task 4

## 1 
Factorize n = 275621053. You can assume that n = pq, where p−q is relatively small. Show
your calculation steps.


Since n-q is relatively small, these numbers must be close to each other. That means that they must also be close to the square root of n. I will start at sqrt(n) and go upwards/downwards on p/q (based on the product compared to n), and find two numbers that multiply to n

In [3]:
import math

n = 275621053
p, q = int(math.sqrt(n)), int(math.sqrt(n))
while p * q != n:
  if p * q < n:
    p += 1
  else:
    q -= 1
print("n =", n)
print("p =", p, "q =", q)
print("p * q =", p * q)

n = 275621053
p = 17021 q = 16193
p * q = 275621053


## Task 2

### a
Alice wants to set up her RSA encryption with private key (n, d) with n = pq, using two
primes p and q, and private key d = 3. She chooses p = 1283, but wonders which of the
following choices for q she should use (NB! They are all prime numbers):
$$ 1307, 1879, 2003, 2027 $$

Explain why she should use q = 2027 for the system to work and to be most secure. For
the weak choices of q, name an effective attack to factorize n (of course, these numbers
are far too small to be secure, so consider the security in relative terms.) <hr>


Since p = 1283, we want a number that is not so close to this. This means that 1307 is not an option. We then look at the other options for q. Both  (p-1) and (q-1) should not have small factors (other than 2 obviously). This is so that we make it more secure against Pollard's p-1 algorithm. We look at the factors for q-1:

In [9]:
def print_factors(x):
   print("The factors of",x,"are:")
   for i in range(2, int(x/2 + 1)):
       if x % i == 0:
           print(i)

numbers = [1879, 2003, 2027]
for number in numbers:
  print_factors(number-1)

The factors of 1878 are:
2
3
6
313
626
939
The factors of 2002 are:
2
7
11
13
14
22
26
77
91
143
154
182
286
1001
The factors of 2026 are:
2
1013


Here we see that 2026 (2027 - 1) is 2*1013, which is a prime multiplied with 2. This makes 2027 a good choice for q, since it's also not so similar to q (1283)

### b
Find the corresponding public key e using the extended Euclidean algorithm. Write a
program to do the calculation.<hr>